### PR TITLE
Support LOITER_TO_ALT in Copter

### PR DESCRIPTION
**Pull request by peterbarker**
Uses waypoint navigation to get to any specified destination, then uses loiter controller to correct altitude.

If precland (precloiter) is in play then it will be used.

### CODE REVIEW

File: ArduCopter/mode_auto.cpp [hunk 1]
```diff
@@ -933,6 +944,64 @@ void Copter::ModeAuto::loiter_run()
...
+            return;
+        }
+    }
+
+    if (!loiter_to_alt.loiter_start_done) {
+        loiter_nav->init_target();
+        _mode = Auto_LoiterToAlt;
+        loiter_to_alt.loiter_start_done = true;
+    }
+    const float current_alt_cm = inertial_nav.get_altitude();
```
Comment (khancyr) [inline ArduCopter/mode_auto.cpp]: float current_alt_cm;
if (ahrs.get_relative_position_D_origin(current_alt_cm)) {
   current_alt_cm = -current_alt_cm * 100;  // convert from m in NED to cm in NEU
}

Comment (khancyr) [reply ArduCopter/mode_auto.cpp]: or simply copter.current_loc.alt

File: ArduCopter/mode.h [hunk 2]
```diff
@@ -398,6 +401,14 @@ class ModeAuto : public Mode {
     uint16_t loiter_time_max;                // How long we should stay in Loiter Mode for mission scripting (time in seconds)
     uint32_t loiter_time;                    // How long have we been loitering - The start time in millis
 
+    struct {
+        bool reached_destination_xy : 1;
```
Comment (khancyr) [inline ArduCopter/mode.h]: bool \o/

File: ArduCopter/mode_auto.cpp [hunk 3]
```diff
@@ -933,6 +944,64 @@ void Copter::ModeAuto::loiter_run()
...
+        loiter_to_alt.reached_alt = true;
+    } else if (alt_error_cm * loiter_to_alt.alt_error_cm < 0) {
+        // we were above and are now below, or vice-versa
+        loiter_to_alt.reached_alt = true;
+    }
+    loiter_to_alt.alt_error_cm = alt_error_cm;
+
+    // loiter...
+
+    land_run_horizontal_control();
```
Comment (khancyr) [inline ArduCopter/mode_auto.cpp]: precision_loiter_xy() ?

Comment (khancyr) [reply ArduCopter/mode_auto.cpp]: just copy Loiter_Flying not landing stuff ...

**Comment by rmackay9** [comment]
I wonder if we can perhaps get away without the reached_wp_destination_xy.  It depends upon what behaviour we want but if we go for a behaviour in which the vehicle first flies to the specified lat/lon and then climbs or decends to the target altitude then we could implement it much like we implement a LAND.  The issue would be that the vertical loiter/land controller is rate based so it does not allow entering a target altitude.

**Comment by khancyr** [comment]
there is the same problem in https://github.com/ArduPilot/ardupilot/pull/8932 . 
Maybe we should first make submode for Auto as we speak with @peterbarker

File: ArduCopter/mode_auto.cpp (hunk 3 above)
Comment (khancyr) [reply ArduCopter/mode_auto.cpp]: well land_run_horizontal_control is not well named ... but that is good ...

Comment (peterbarker) [reply ArduCopter/mode_auto.cpp]: I assume you're OK with the current code, then?!

**Comment by peterbarker** [comment]
@rmackay9 I'd obviously assumed you'd wp-navigate to the loiter point then make any altitude changes required using the loiter/precision loiter code.

That was based on the fact we do that for precision landing and for landing.

The current code does the z stuff by itself when in the loiter phase.

If we do without the `reached_wp_destination_xy()` then the vehicle will potentially climb to within the waypoint radius using the waypoint controller - in which case you may as well just use a navigation waypoint :-)

**Comment by peterbarker** [comment]
I've removed the accel-imiting in thez-controller from this PR.  It probably wasn't right, and is a pre-existing issue.

**Comment by rmackay9** [comment]
This looks good to me.. I think I'll test it in SITL and if it's OK then I'll merge it.

**Comment by rmackay9** [comment]
Do we need to add this to these two places as well?

1. AP_Mission::Mission_Command::type()
2. AP_Terrain::update_mission_data()

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**